docs/source/advanced_patterns.rst

```diff
@@ -0,0 +1,107 @@
+.. _advanced_communication_patterns:
+
+
+*******************************
+Advanced communication patterns
+*******************************
+
+Some advanced communication patterns are also implemented in osBrain. They are
+called channels, rather than sockets, as they are formed with multiple sockets.
+
+
+.. index:: async-req-rep
+.. _async_req_rep:
+
+Asynchronous Request-Reply channel
+==================================
+
+An asynchronous request-reply channel is very similar to a normal request-reply
+pattern, only in this case we want to process the reply asynchronously (i.e.:
+assign a handler for this reply and forget about it until it is received).
+
+This means that, when using an asynchronous request-reply channel, a handler
+must be specified not only for the replier but also for the requester.
+
+Example
+-------
+
+See the following example in which Bob sends an asynchronous request to Alice.
+Note that:
+
+- When Alice binds, she uses ``ASYNC_REP`` instead of ``REP``.
+- Bob is assigned a handler too to process Alice reply when it is received.
+- Bob sends the request and is automatically freed. It can log messages or do
+  other stuff.
+- When the reply is received, Bob automatically processes it.
+
+.. literalinclude:: ../../examples/async_req_rep.py
+
+That is the most basic example, but channels are a bit more flexible. When
+sending an asynchronous request we may want to manually specify a handler for
+the reply. This will overwrite the default handler specified when connecting to
+the ``ASYNC_REP`` server. Try to change the ``.send()`` call::
+
+   def deaf(agent, message):
+       agent.log_info('I am deaf...')
+
+
+   bob.send('alice', 'Hello, Alice!', handler=deaf)
+
+We can also specify a maximum wait time for the reply and some code to be
+executed in case the reply was not received after that wait. Try the example
+above, but setting ``wait=0.5``::
+
+   def no_reply_in_time(agent):
+       agent.log_warning('No reply received!')
+
+
+   bob.send('alice', 'Hello, Alice!', wait=0.5, on_error=no_reply_in_time)
+
+.. note:: If ``on_error`` is not specified, by default the agent will simply
+   log a warning.
+
+
+.. index:: sync-pub-sub
+.. _sync_pub_sub:
+
+Synced Publish-Subscribe channel
+================================
+
+Another common use case is the synced publish-subscribe channel. It is similar
+to the normal publish-subscribe pattern but has some extra functionality. The
+server is configured to publish messages through a ``PUB`` socket and clients
+can connect to it to receive these messages. Using the synced publish-subscribe
+channel, however, allows the clients to send requests to the publisher.
+
+This can be really useful when we want to share some state/data from the
```

Comment:
What do you mean by "state/data"? If by "state" you mean the current status of the agent isn't that "data" after all?

Maybe simplify this to just "data" or "information" (the same goes for the sentence below).